src/schema/ecommerce/approve_order_mutation.ts

```diff
@@ -0,0 +1,80 @@
+import { graphql } from "graphql"
+import { OrderReturnType } from "schema/ecommerce/types/order_return"
+import { OrderMutationInputType } from "schema/ecommerce/types/order_mutation_input"
+import { mutationWithClientMutationId } from "graphql-relay"
+import {
+  RequestedFulfillmentFragment,
+  BuyerSellerFields,
+} from "./query_helpers"
+import gql from "lib/gql"
+import { OrderOrFailureUnionType } from "./types/order_or_error_union"
+
+export const ApproveOrderMutation = mutationWithClientMutationId({
+  name: "ApproveOrder",
+  description: "Approves an order with payment",
+  inputFields: OrderMutationInputType.getFields(),
+  outputFields: {
+    orderOrError: {
+      type: OrderOrFailureUnionType,
+    },
+  },
+  mutateAndGetPayload: (
+    { orderId },
+    context,
+    { rootValue: { accessToken, exchangeSchema } }
+  ) => {
+    if (!accessToken) {
+      return new Error("You need to be signed in to perform this action")
+    }
+    const mutation = gql`
+      mutation approveOrder($orderId: ID!) {
+        ecommerce_approveOrder(input: {
+          id: $orderId,
+        }) {
+          orderOrError {
+            ... on EcommerceOrderWithMutationSuccess {
+              order {
+              id
+                code
+                currencyCode
+                state
+                ${BuyerSellerFields}
+                ${RequestedFulfillmentFragment}
+                itemsTotalCents
+                shippingTotalCents
+                taxTotalCents
+                commissionFeeCents
+                transactionFeeCents
+                buyerTotalCents
+                sellerTotalCents
+                updatedAt
+                createdAt
+                stateUpdatedAt
+                stateExpiresAt
+                lineItems{
+                  edges{
+                    node{
+                      id
+                      priceCents
+                      artworkId
+                      editionSetId
+                      quantity
+                    }
+                  }
+                }
+              }
+            }
+            ... on EcommerceOrderWithMutationFailure {
+              error {
+                description
+              }
+            }
+          }
+        }
+      }
+    `
+    return graphql(exchangeSchema, mutation, null, context, {
+      orderId,
+    }).then(result => result.data!.ecommerce_approveOrder)
```

Comment:
Before these callbacks were inspecting `result.errors` which is ignored now. Is that a potential source of protocol-level error messages that we should be propagating?

Reply:
i _think_ that was for us to be able to read response in old format from Exchange and properly raise errors here, i think in this current case we are fine mainly since this call basically passes Exchanges response up and if there are errors it will be in the response returned from MP.

Reply:
This is were I added those: https://github.com/artsy/metaphysics/pull/1209

I don't exactly remember the specific query I was running but I think it was something like confirming an order using authentication token from a wrong partner. @ashkan18 it would be great to give it a try because it seems this kind of errors bypass exchange error handling, unless that has been updated too.

Thanks @ds300 for catching it.

Reply:
i _think_ we are good here now since this basically gets the response from exchange (either Order or an error) and returns it, before we had to see if there were errors and raise them to get MP style errors i think, but lets keep an eye on it. Im never 100% sure here :)

Reply:
@alloy and I were debugging the issue described here: https://artsy.slack.com/archives/C1HH3KNJG/p1535634594000100 and discovered that these `results.errors` are indeed important. We were seeing a 401 error manifesting as a `cannot set property 'clientMutationId' of null` error because the `result.data.ecommerce_<foo>` properties are null when exchange rejected the request. So I'll make a PR to add the checks back in.

Reply:
awesome! thanks! sorry for causing it, i guess here we'd need to check for `result.data.ecommerce_...orderOrError.error`?

Reply:
That one gets propagated to the client, we'd need to check for `result.errors` here, just like it was doing before.

Reply:
`result.data.ecommerce_...orderOrError.error` are user errors, other uncaught (runtime/protocol) errors thrown during execution will still go to the top-level errors.